src/QsFmt/Formatter/Rules.fs

```diff
@@ -101,3 +101,73 @@ let newLines =
             else
                 { block with CloseBrace = Terminal.mapPrefix ensureNewLine block.CloseBrace }
     }
+
+/// <summary>
+/// Gets the trivia from a terminal option.
+/// </summary>
+let getTrivia paren =
+    match paren with
+    | None -> []
+    | Some p -> p.Prefix
+
+let qubitBindingUpdate =
+    { new Rewriter<Trivia list>() with
```

Comment:
The trivia list is the trivia from the removed parentheses around the qubit binding expression. These parentheses get removed, but we need to keep the trivia from them and append them to the appropriate next terminal. This is done by selectively passing the trivia list through the context. Do you have a different way of doing this in mind?

Reply:
Oh, I missed that, thanks.

The context is meant to be used for information that applies to all children of a particular node, so using it for something that should apply only to the left-most child is a little risky (for example if the QubitBinding or Tuple handlers are changed to recurse into the other children). For this I'd recommend using one of the `mapPrefix` functions - example:

https://github.com/microsoft/qsharp-compiler/blob/6d20b62f8822044974e3ae0027df910158454bd4/src/QsFmt/Formatter/SyntaxTree/Statement.fs#L45

If there isn't one defined for the node type you want, you can define a new one.

Reply:
It's a little difficult for me to understand just from the code what these functions do and how to use them. Could you give me some more detail on these `mapPrefix` functions?

Reply:
In general, `N.mapPrefix` says, "for some node type `N`, find its left-most terminal child and apply a mapping function to its trivia prefix." An example of how to use it is the indentation rewriter:

https://github.com/microsoft/qsharp-compiler/blob/6d20b62f8822044974e3ae0027df910158454bd4/src/QsFmt/Formatter/Rules.fs#L63-L76

It basically is just a convenient way to do something to the prefix of a branch node when you don't know/care about how many steps it takes to reach its terminal nodes.

Reply:
Okay so the mapPrefix takes a mapper function and applies that function to the Trivia list of the leftmost Terminal node of the corresponding data structure. Currently the defined mapPrefix functions are on the NamespaceItem, Statement, and of course Terminal types. Have I got that right? I'm not sure how passing a mapping function around gets the effect I need on this. I need to remember the Trivia list from the Terminal I'm erasing to give it to the next Terminal. I can see the indentation rewriter is doing something like that with `level` but I don't understand how/what it is doing.

Reply:
For example, instead of

```fsharp
{ ``use`` with
    // ...
    Semicolon = rewriter.Terminal(closeTrivia, ``use``.Semicolon)
}
```

try

```fsharp
{ ``use`` with
    // ...
    Semicolon = Terminal.mapPrefix ((@) closeTrivia) ``use``.Semicolon
}
```

You can do the same thing for Binding and Block (you may need to define new `mapPrefix` functions for them). I'm pretty sure you should be able to write this transformation without any context (i.e., `qubitBindingUpdate` can be of type `unit Rewriter`).

Reply:
At levels higher than Terminal, if I use a mapPrefix function instead of the rewriter call, the rewriter won't traverse the tree lower than that point. This becomes an issue, especially, with the Block nodes.

Reply:
Did you try recursing before/after (not sure which is better) updating the prefix, like

```fsharp
rewriter.Block(context, Block.mapPrefix ((@) closeTrivia) ``use``.Block)
```

or

```fsharp
rewriter.Block(context, ``use``.Block) |> Block.mapPrefix ((@) closeTrivia)
```

Reply:
Okay I think I have something that works with the mapPrefix functions now. Let me know if this is what you were thinking.